File: clients/cpp/model/Loot.cpp
#include "Loot.hpp"

namespace model {

Loot::Loot(int id, model::Vec2 position, model::Item item) : id(id), position(position), item(item) { }

// Read Loot from input stream
Loot Loot::readFrom(InputStream& stream) {
    int id = stream.readInt();
    model::Vec2 position = model::Vec2::readFrom(stream);
    model::Item item = model::readItem(stream);
    return Loot(id, position, item);
}

// Write Loot to output stream
void Loot::writeTo(OutputStream& stream) const {
    stream.write(id);
    position.writeTo(stream);
    writeItem(item, stream);
}

// Get string representation of Loot
std::string Loot::toString() const {
    std::stringstream ss;
    ss << "Loot { ";
    ss << "id: ";
    ss << id;
    ss << ", ";
    ss << "position: ";
    ss << position.toString();
    ss << ", ";
    ss << "item: ";
    ss << itemToString(item);
    ss << " }";
    return ss.str();
}

}

File: clients/go/model/loot.go
package model

import (
	"fmt"
	"io"

	. "aicup22/stream"
)

// Loot lying on the ground
type Loot struct {
	// Unique id
	Id int32
	// Position
	Position Vec2
	// Item
	Item Item
}

func NewLoot(id int32, position Vec2, item Item) Loot {
	return Loot{
		Id:       id,
		Position: position,
		Item:     item,
	}
}

// Read Loot from reader
func ReadLoot(reader io.Reader) Loot {
	var id int32
	id = ReadInt32(reader)
	var position Vec2
	position = ReadVec2(reader)
	var item Item
	item = ReadItem(reader)
	return Loot{
		Id:       id,
		Position: position,
		Item:     item,
	}
}

// Write Loot to writer
func (loot Loot) Write(writer io.Writer) {
	id := loot.Id
	WriteInt32(writer, id)
	position := loot.Position
	position.Write(writer)
	item := loot.Item
	item.Write(writer)
}

// Get string representation of Loot
func (loot Loot) String() string {
	stringResult := "{ "
	stringResult += "Id: "
	id := loot.Id
	stringResult += fmt.Sprint(id)
	stringResult += ", "
	stringResult += "Position: "
	position := loot.Position
	stringResult += position.String()
	stringResult += ", "
	stringResult += "Item: "
	item := loot.Item
	stringResult += item.String()
	stringResult += " }"
	return stringResult
}
